deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/layers/DepthwiseConvolution2D.java

```diff
@@ -0,0 +1,178 @@
+        this.dilation = builder.dilation;
+        if (builder.kernelSize.length != 2)
+            throw new IllegalArgumentException("Kernel size of should be rows x columns (a 2d array)");
+        this.kernelSize = builder.kernelSize;
+        if (builder.stride.length != 2)
+            throw new IllegalArgumentException("Stride should include stride for rows and columns (a 2d array)");
+        this.stride = builder.stride;
+        if (builder.padding.length != 2)
+            throw new IllegalArgumentException("Padding should include padding for rows and columns (a 2d array)");
+        this.padding = builder.padding;
+        this.cudnnAlgoMode = builder.cudnnAlgoMode;
+        this.cudnnFwdAlgo = builder.cudnnFwdAlgo;
+        this.cudnnBwdFilterAlgo = builder.cudnnBwdFilterAlgo;
+        this.cudnnBwdDataAlgo = builder.cudnnBwdDataAlgo;
+        this.nOut = this.nIn * this.depthMultiplier;
+
+        initializeConstraints(builder);
+    }
+
+    public boolean hasBias() {
+        return hasBias;
+    }
+
+    @Override
+    public DepthwiseConvolution2D clone() {
+        DepthwiseConvolution2D clone = (DepthwiseConvolution2D) super.clone();
+        if (clone.kernelSize != null)
+            clone.kernelSize = clone.kernelSize.clone();
+        if (clone.stride != null)
+            clone.stride = clone.stride.clone();
+        if (clone.padding != null)
+            clone.padding = clone.padding.clone();
+        return clone;
+    }
+
+
+    @Override
+    public Layer instantiate(NeuralNetConfiguration conf, Collection<TrainingListener> trainingListeners,
+                             int layerIndex, INDArray layerParamsView, boolean initializeParams) {
+        LayerValidation.assertNInNOutSet(
+                "DepthwiseConvolution2D", getLayerName(), layerIndex, getNIn(), getNOut());
+
+        DepthwiseConvolution2DLayer ret = new DepthwiseConvolution2DLayer(conf);
+        ret.setListeners(trainingListeners);
+        ret.setIndex(layerIndex);
+        ret.setParamsViewArray(layerParamsView);
+        Map<String, INDArray> paramTable = initializer().init(conf, layerParamsView, initializeParams);
+        ret.setParamTable(paramTable);
+        ret.setConf(conf);
+
+        return ret;
+    }
+
+    @Override
+    public ParamInitializer initializer() {
+        return DepthwiseConvolutionParamInitializer.getInstance();
+    }
+
+    @Override
+    public InputType getOutputType(int layerIndex, InputType inputType) {
+        if (inputType == null || inputType.getType() != InputType.Type.CNN) {
+            throw new IllegalStateException("Invalid input for  depth-wise convolution layer (layer name=\""
+                    + getLayerName() + "\"): Expected CNN input, got " + inputType);
+        }
+
+
+        return InputTypeUtil.getOutputTypeCnnLayers(inputType, kernelSize, stride, padding, dilation,
+                convolutionMode, nOut, layerIndex, getLayerName(), DepthwiseConvolution2DLayer.class);
```

Comment:
This might not be correct, unless nOut is set to the correct value elsewhere?

Reply:
that's set correctly before, it's arguably the one thing one has to really be careful about with depth conv, so thanks for flagging this.